File: sdks/typescript/src/http/HttpClient.ts
import { DevKitError, AuthenticationError } from '../models/errors';

export class HttpClient {
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly timeout: number;

  constructor(baseUrl: string, apiKey: string, timeout: number = 10000) {
    this.baseUrl = baseUrl.endsWith('/') ? baseUrl.slice(0, -1) : baseUrl;
    this.apiKey = apiKey;
    this.timeout = timeout;
  }

  private async request<T>(
    path: string,
    options?: RequestInit
  ): Promise<T> {
    const url = `${this.baseUrl}${path}`;
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await fetch(url, {
        ...options,
        signal: controller.signal,
        headers: {
          'Authorization': `Bearer ${this.apiKey}`,
          'Content-Type': 'application/json',
          ...options?.headers,
        },
      });

      clearTimeout(timeoutId);

      if (response.status === 401) {
        throw new AuthenticationError('Invalid API key');
      }

      if (!response.ok) {
        throw new DevKitError(`HTTP error: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      if (error instanceof DevKitError) {
        throw error;
      }
      if (error instanceof Error && error.name === 'AbortError') {
        throw new DevKitError('Request timeout');
      }
      throw new DevKitError('Network error', error as Error);
    }
  }

  async get<T>(path: string): Promise<T> {
    return this.request<T>(path, { method: 'GET' });
  }

  async post<T>(path: string, body?: unknown): Promise<T> {
    return this.request<T>(path, {
      method: 'POST',
      body: body ? JSON.stringify(body) : undefined,
    });
  }

  async put<T>(path: string, body?: unknown): Promise<T> {
    return this.request<T>(path, {
      method: 'PUT',
      body: body ? JSON.stringify(body) : undefined,
    });
  }

  async delete(path: string): Promise<void> {
    await this.request<void>(path, { method: 'DELETE' });
  }
}

File: sdks/flutter/lib/http/http_client.dart
import 'dart:async';
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/errors.dart';

class HttpClient {
  final String baseUrl;
  final String apiKey;
  final Duration timeout;

  HttpClient({
    required String baseUrl,
    required this.apiKey,
    this.timeout = const Duration(milliseconds: 10000),
  }) : baseUrl = baseUrl.endsWith('/')
            ? baseUrl.substring(0, baseUrl.length - 1)
            : baseUrl;

  Future<T> _request<T>(
    String path, {
    String method = 'GET',
    Object? body,
  }) async {
    final url = '$baseUrl$path';
    final headers = {
      'Authorization': 'Bearer $apiKey',
      'Content-Type': 'application/json',
    };

    late http.Response response;
    try {
      if (method == 'GET') {
        response =
            await http.get(Uri.parse(url), headers: headers).timeout(timeout);
      } else if (method == 'POST') {
        response = await http
            .post(
              Uri.parse(url),
              headers: headers,
              body: body != null ? jsonEncode(body) : null,
            )
            .timeout(timeout);
      } else if (method == 'PUT') {
        response = await http
            .put(
              Uri.parse(url),
              headers: headers,
              body: body != null ? jsonEncode(body) : null,
            )
            .timeout(timeout);
      } else if (method == 'DELETE') {
        response = await http
            .delete(Uri.parse(url), headers: headers)
            .timeout(timeout);
      } else {
        throw DevKitError('Unsupported HTTP method: $method');
      }

      if (response.statusCode == 401) {
        throw AuthenticationError('Invalid API key');
      }

      if (response.statusCode >= 400) {
        throw DevKitError('HTTP error: ${response.statusCode}');
      }

      if (response.statusCode == 204) {
        return null as T;
      }

      final responseBody =
          response.body.isNotEmpty ? jsonDecode(response.body) : null;

      return responseBody as T;
    } on TimeoutException {
      throw DevKitError('Request timeout');
    } on AuthenticationError {
      rethrow;
    } on DevKitError {
      rethrow;
    } catch (e) {
      throw DevKitError('Network error', e);
    }
  }

  Future<T> get<T>(String path) => _request<T>(path);

  Future<T> post<T>(String path, [Object? body]) =>
      _request<T>(path, method: 'POST', body: body);

  Future<T> put<T>(String path, [Object? body]) =>
      _request<T>(path, method: 'PUT', body: body);

  Future<void> delete(String path) => _request<void>(path, method: 'DELETE');
}
